janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java

```diff
@@ -498,58 +511,96 @@ private static Query numericQuery(String key, Cmp relation, Number value) {
         }
     }
 
-    private SearchParams convertQuery(Condition<?> condition, KeyInformation.StoreRetriever information) {
-        SearchParams params = new SearchParams();
+    // adapted from SolrIndex
+    private List<String> customTokenize(Analyzer analyzer, String fieldName, String value) {
+        final List<String> terms = new ArrayList<>();
+        try (CachingTokenFilter stream = new CachingTokenFilter(analyzer.tokenStream(fieldName, value))) {
+            final TermToBytesRefAttribute termAtt = stream.getAttribute(TermToBytesRefAttribute.class);
+            stream.reset();
+            while (stream.incrementToken()) {
+                terms.add(termAtt.getBytesRef().utf8ToString());
+            }
+            return terms;
+        } catch (IOException e) {
+            throw new IllegalArgumentException(e.getMessage(), e);
+        }
+    }
+
+    private void tokenize(SearchParams params, final Mapping mapping, final LuceneCustomAnalyzer delegatingAnalyzer, String value, String key, JanusGraphPredicate janusgraphPredicate) {
+        final Analyzer analyzer = delegatingAnalyzer.getWrappedAnalyzer(key);
+        final List<String> terms = customTokenize(analyzer, key, value);
+        if (terms.size() == 1) {
+            if (janusgraphPredicate == Cmp.EQUAL || janusgraphPredicate == Text.CONTAINS) {
+                params.addQuery(new TermQuery(new Term(key, value)));
+            } else if (janusgraphPredicate == Cmp.NOT_EQUAL) {
+                final BooleanQuery.Builder q = new BooleanQuery.Builder();
+                q.add(new MatchAllDocsQuery(), BooleanClause.Occur.MUST);
+                q.add(new TermQuery(new Term(key, value)), BooleanClause.Occur.MUST_NOT);
+                params.addQuery(q.build(), BooleanClause.Occur.MUST);
+            } else if (janusgraphPredicate == Text.CONTAINS_PREFIX) {
+                final Term term = new Term(key, terms.get(0).toLowerCase());
+                params.addQuery(new PrefixQuery(term), BooleanClause.Occur.MUST);
+            }
```

Comment:
Should there be an `else` here that throws an IllegalArgumentException or something?

Reply:
I'm agree you should throw an IllegalArgumentException and add a test.

Reply:
I added the else clause but, though I'm willing to improve the test coverage, I don't understand how I should proceed for the test :
 
the tokenize method is private and only used in the convertQuery, which is private too, with the predicates Text.CONTAINS, Text.CONTAINS_PREFIX, Cmp.EQUAL and Cmp.NOT_EQUAL.
So, I see no way to check that the Tokenize function throws an IllegalArgumentException (when it uses only one token) with an operator that isn't one of those 4.

Could you provide further guidance, please?

Reply:
Actually, I'm not a professionnal software developer (though I have been doing android software development for the past 6 years, doing solo cowboy coding), I'm a maths teacher... So, I don't have much practice with tests, gits and collaboration (though I understand the value they provide)